### PR TITLE
Remove hooks and consolidate logic in GestureMonitor

### PR DESCRIPTION
**Pull request by j-piasecki**
## Description

 - Removes `useGesture` and `useAnimatedGesture` hooks. Instead of them you can pass `BaseGesture` or `InteractionBuilder` directly to `gesture` (or the new `animatedGesture`)  prop of the `GestureMonitor`.
 - `GestureMonitor` has been refactored and inherited most of the logic from hooks.
 - Examples have been updated to support these changes.
## Test plan

Tested on the Example app

### CODE REVIEW

File: src/handlers/gestures/GestureMonitor.tsx [hunk 1]
```diff
@@ -1,93 +1,399 @@
...
+          } else if (
+            (event.oldState === State.BEGAN ||
+              event.oldState === State.UNDETERMINED) &&
+            event.state === State.ACTIVE
+          ) {
+            if (gesture.isOnStartWorklet) {
+              gesture.onStart?.(event);
+            } else if (gesture.onStart) {
+              console.warn('Animated gesture callback must be a worklet');
+            }
```
Comment (jakub-gonet) [inline src/handlers/gestures/GestureMonitor.tsx:266]: Those ifs should be extracted to helper.

File: src/handlers/gestures/GestureMonitor.tsx [hunk 2]
```diff
@@ -1,93 +1,399 @@
...
+  if (useAnimated) {
+    // Whether animatedGesture or gesture is used shouldn't change
+    // during while an app is running
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    useAnimatedGesture(preparedGesture);
+  }
+
+  useEffect(() => {
+    firstRenderRef.current = true;
+    const viewTag = findNodeHandle(viewRef.current) as number;
```
Comment (jakub-gonet) [inline src/handlers/gestures/GestureMonitor.tsx:348]: Why do you need to cast it here?

File: src/handlers/gestures/GestureMonitor.tsx [hunk 3]
```diff
@@ -1,93 +1,399 @@
...
+      </AnimatedWrap>
+    );
+  } else {
+    return <Wrap ref={viewRef}>{props.children}</Wrap>;
   }
-}
+};
 
-class Wrap extends React.Component<{ onGestureHandlerEvent: unknown }> {
+class Wrap extends React.Component<{ onGestureHandlerEvent?: unknown }> {
```
Comment (jakub-gonet) [inline src/handlers/gestures/GestureMonitor.tsx:396]: Is it unknown though?

File: src/handlers/gestures/GestureMonitor.tsx [hunk 4]
```diff
@@ -1,93 +1,399 @@
...
+      useAnimated
+    );
+
+    return () => {
+      dropHandlers(preparedGesture);
+    };
+  }, []);
+
+  useEffect(() => {
+    if (!firstRenderRef.current) {
```
Comment (jakub-gonet) [inline src/handlers/gestures/GestureMonitor.tsx:363]: Are you checking if it's falsy or false here?

File: src/handlers/gestures/GestureMonitor.tsx (hunk 2 above)
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx:348]: I used the existing code: https://github.com/software-mansion/react-native-gesture-handler/blob/master/src/handlers/createHandler.ts#L259

File: src/handlers/gestures/GestureMonitor.tsx (hunk 3 above)
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx:396]: In this case it is the result of the `useEvent` hook from `react-native-reanimated` and it's not documented (at least I didn't find anything useful), so kind of?

File: src/handlers/gestures/GestureMonitor.tsx (hunk 4 above)
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx:363]: I'm checking if it's false.

File: examples/Example/src/new_api/calculator/index.tsx [hunk 5]
```diff
@@ -326,24 +321,22 @@ function Button({ text, append }: ButtonProps) {
...
-          runOnJS(append)(text);
-        }
-      })
-      .setOnBegan((_e) => {
-        'worklet';
-        alpha.value = withTiming(0.75, { duration: TAP_ANIMATION_DURATION });
-      })
-  );
+  const tapHandler = Gesture.tap()
+    .setOnEnd((_e, s) => {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/new_api/calculator/index.tsx]: Done in 273ceabbfc34b1616931fa8e564e2a09d9e71fdb

File: src/handlers/gestures/GestureMonitor.tsx [hunk 6]
```diff
@@ -1,93 +1,399 @@
...
+        gesture.handlerTag,
+        viewTag,
+        false
+      );
+    } else {
+      RNGestureHandlerModule.attachGestureHandler(
+        gesture.handlerTag,
+        viewTag,
+        true
+      );
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx]: Done in 5e88430908b2c8c26be9628b25531646d0fccb85

File: src/handlers/gestures/GestureMonitor.tsx [hunk 7]
```diff
@@ -1,93 +1,399 @@
...
+
+    setImmediate(() => {
+      gestureConfig?.prepare();
+
+      let requireToFail: number[] = [];
+      if (handler.config.requireToFail) {
+        requireToFail = handler.config.requireToFail
+          .map(convertToHandlerTag)
+          .filter((tag) => tag > 0);
+      }
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx]: Done in 9f1b973c5adf8df2262a6038078200e349e3ed9c

File: src/handlers/gestures/GestureMonitor.tsx [hunk 8]
```diff
@@ -1,93 +1,399 @@
...
-      if (typeof ref === 'function') {
-        ref(node);
-      } else {
-        ref.current = node;
-      }
+  return false;
+}
+
+function useAnimatedGesture(preparedGesture: GestureConfigReference) {
+  if (!Reanimated) return;
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx]: Fixed in 860cfc2c3a8a90cf15f829ccdf79676f08dede29

File: src/handlers/gestures/GestureMonitor.tsx [hunk 9]
```diff
@@ -1,93 +1,399 @@
...
+  for (const handler of gesture) {
+    RNGestureHandlerModule.createGestureHandler(
+      handler.handlerName,
+      handler.handlerTag,
+      filterConfig(handler.config, ALLOWED_PROPS)
+    );
+
+    registerHandler(handler.handlerTag, handler);
+
+    setImmediate(() => {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx:109]: Done in e593ee19a3d509b76e73586e9ad21ed53bd7ff49

File: src/handlers/gestures/GestureMonitor.tsx [hunk 10]
```diff
@@ -1,93 +1,399 @@
...
-        this.attachGestureHandlers(
-          RNRenderer.findHostInstance_DEPRECATED(this)._nativeTag
-        );
-      };
+function attachHandlers(
+  preparedGesture: GestureConfigReference,
+  gestureConfig: InteractionBuilder | GestureType | undefined,
+  gesture: GestureType[],
+  viewTag: number,
+  useAnimated: boolean
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx]: Done in c740f753e6c331a34c1e59f9c147414c63cffcc5

File: src/handlers/gestures/GestureMonitor.tsx [hunk 11]
```diff
@@ -1,93 +1,399 @@
...
+    for (let i = 0; i < currentCallback.length; i++) {
+      const gesture = currentCallback[i];
+
+      if (event.handlerTag === gesture.handlerTag) {
+        if (isStateChangeEvent(event)) {
+          if (
+            event.oldState === State.UNDETERMINED &&
+            event.state === State.BEGAN
+          ) {
+            if (gesture.isOnBeganWorklet) {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx]: Done in e3c6c64c724342f61b35c2b02a2688ac3b0af2f4

File: src/handlers/gestures/GestureMonitor.tsx (hunk 1 above)
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx]: Done in e3c6c64c724342f61b35c2b02a2688ac3b0af2f4

File: src/handlers/gestures/GestureMonitor.tsx [hunk 12]
```diff
@@ -1,93 +1,399 @@
...
+import { forceTouchGestureHandlerProps } from '../ForceTouchGestureHandler';
+import { longPressGestureHandlerProps } from '../LongPressGestureHandler';
+import {
+  panGestureHandlerProps,
+  panGestureHandlerCustomNativeProps,
+} from '../PanGestureHandler';
+import { tapGestureHandlerProps } from '../TapGestureHandler';
+import { State } from '../../State';
 
 startListening();
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx]: Something like in https://github.com/software-mansion/react-native-gesture-handler/pull/1595/commits/7b28aec1e341800dc4e48e45fe1bcb45f75a547d?

[the remark this answers is not in the document]
Comment (jakub-gonet) [reply src/handlers/gestures/GestureMonitor.tsx]: Looks great.

**Review by jakub-gonet — COMMENTED** [review]
I think we can add `ContinuousBaseGesture` to not duplicate `onUpdate`.

File: src/handlers/gestures/GestureMonitor.tsx (hunk 3 above)
Comment (jakub-gonet) [reply src/handlers/gestures/GestureMonitor.tsx:420]: https://github.com/software-mansion/react-native-reanimated/blob/3f1ac80d95fbf2736e3747a219a9e2afe133a7a7/src/reanimated2/hook/utils.ts#L14-L18

It's typed but we can leave it for now.

File: src/handlers/gestures/gesture.ts [hunk 13]
```diff
@@ -102,8 +111,9 @@ export abstract class BaseGesture<
     ) => void
   ) {
     this.handlers.onBegan = callback;
-    //@ts-ignore if callback is a worklet, the property will be available, if not then the check will return false
-    this.handlers.isOnBeganWorklet = callback.__workletHash != null;
+    this.handlers.isWorklet[CALLBACK_TYPE.BEGAN] =
+      //@ts-ignore if callback is a worklet, the property will be available, if not then the check will return false
+      callback.__workletHash != null;
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/gesture.ts:116]: Done in 6bdba900217dd9e03a30078efafaa9363daf4156

[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/gesture.ts:116]: `__workletHash` is null when pass normal functions as callbacks. Either you don't use `animatedGesture` or you forgot to mark them as `worklet`.

[the remark this answers is not in the document]
Comment (jakub-gonet) [reply src/handlers/gestures/gesture.ts:116]: `null` or `undefined`?
I'd expect it's null so we can make this check more precise: `callback.__workletHash !== undefined`

[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/gesture.ts]: You're right, it's `undefined`. I updated it in 24d95f45e3239231ceaa9be9764a8cdca488950e.

**Review by jakub-gonet — APPROVED** [review]
:shipit:

**Pull request merged**